source/Tool/CorpusImpl.cpp

```diff
@@ -185,11 +185,109 @@ build(
+        {
+            llvm::StringRef rf{ns};
+            std::size_t idx = rf.find("::"), last = 0u;
+            for (;
+                    idx != std::string::npos;
+                    idx = rf.find("::", (last = (idx + 2))))
+                filterSegments.push_back(rf.substr(last, idx - last));
+            filterSegments.push_back(rf.substr(last));
+        }
+
+        if (filterSegments.empty())
+            continue;
+        // the reason the absolute thing works is because of the global namespace
+        for (const auto & im : InfoMap)
+        {
+            if (!im.second || !im.second->isNamespace())
+                continue;
+
+            if (im.second->Name == filterSegments.back())
+            {
+                auto fitr = std::next(filterSegments.rbegin());
+                auto &cns = im.second->Namespace;
+                auto citr = cns.begin();
+
+                while (fitr != filterSegments.rend()
+                       && citr != cns.end())
+                {
+                    auto nm = find(*citr);
+                    if (!nm || nm->Name != *fitr)
+                        break;
+                    fitr++;
+                    citr++;
+                }
+
+                if (fitr == filterSegments.rend()) // did match
+                {
+                    auto & f = filters.emplace_back();
+                    f.push_back(im.second->id);
+                    f.insert(f.end(), im.second->Namespace.begin(), im.second->Namespace.end());
+                }
+            }
+        }
+    }
+
+    for (auto & im : InfoMap)
+    {
+        for (auto & filter : filters)
+        {
+            bool needs_removal = false;
+            if (im.second->isNamespace())
+                needs_removal = (filter.size() <= (im.second->Namespace.size() + 1))
+                                && (im.second->id == filter.front())
+                                && std::equal(std::next(filter.begin()), filter.end(), im.second->Namespace.begin());
+            else
+                needs_removal = (filter.size() <= im.second->Namespace.size())
+                    && std::equal(filter.begin(), filter.end(), im.second->Namespace.begin());
+
+            if (!needs_removal)
+                continue;
+
+            if (!im.second->Namespace.empty())
+            {
+                auto parent = im.second->Namespace.front();
+                auto info = find(parent);
+                std::vector<SymbolID> *Members = nullptr;
+                if (info != nullptr)
+                {
+                    if (info->isNamespace())
+                        Members = &static_cast<NamespaceInfo *>(info)->Members;
+                    else if (info->isRecord())
+                        Members = &static_cast<RecordInfo *>(info)->Members;
+                }
+                if (Members)
+                    std::erase(*Members, im.second->id);
+                InfoMap.erase(im.first());
```

Comment:
no you can't do this. the filtering needs to happen during top-level visitation. and even so,  sometimes things can slip through. For example consider the following:
```
namespace detail {
struct Q {};
};

struct P : detail::Q {};
```

In this scenario we should filter `detail::Q` when we see it at the top level during visitation. But when we extract "P", we cannot avoid extracting base classes unless they are private. So then we have to extract `detail::Q` even though it would fail the filter.

You are also not generally allowed to modify the InfoMap after it is created. This all has to be done in ASTVisitor.cpp